### PR TITLE
Add `__match_args__` to node types

### LINKED ISSUE #13243: Add `__match_args__` to Node types
I am working on a project which makes heavy use of Mypy AST trees/nodes. and it would be nice if there was less boilerplate when doing pattern matching against nodes (especially for more complex scenarios).

For example, parsing the following expression:

```python
1 + 2
```

Results in an `OpExpr` expression node (ignoring setup, see below), which I can then pattern match like so:

```python
match o:
    case OpExpr(
        op="+",
        left=IntExpr(value=1),
        right=IntExpr(value=2),
    ):
        print("found")

    case _:
        print("not found")
```

This works, but what would be really nice is being able to do this:

```python
match o:
    case OpExpr("+", IntExpr(1), IntExpr(2)):
        print("found")

    case _:
        print("not found")
```

But doing so causes the following exception:

```
Traceback (most recent call last):
  File "[redacted]/main.py", line 39, in <module>
    tree.accept(Visitor())
  File "mypy/nodes.py", line 347, in accept
  File "mypy/traverser.py", line 42, in visit_mypy_file
  File "mypy/nodes.py", line 1128, in accept
  File "mypy/traverser.py", line 89, in visit_expression_stmt
  File "mypy/nodes.py", line 1888, in accept
  File "[redacted]/main.py", line 30, in visit_op_expr
    case OpExpr(IntExpr(1), "+", IntExpr(2)):
TypeError: OpExpr() accepts 0 positional sub-patterns (3 given)
```

All that would need to done to fix this (I think) is the following:

```diff
 class OpExpr(Expression):
     """Binary operation (other than . or [] or comparison operators,
     which have specific nodes)."""
 
     __slots__ = ('op', 'left', 'right',
                  'method_type', 'right_always', 'right_unreachable')
+
+    __match_args__ ('op', 'left', 'right')
```

And then do that for every node we would (feasibly) pattern match against. I am willing to make a PR for this, assuming all that needs to be done is add a `__match_args__` line to a few classes.

<details>
<summary>Full code example</summary>

```python
from mypy.build import build
from mypy.main import process_options
from mypy.nodes import IntExpr, OpExpr
from mypy.traverser import TraverserVisitor

files, opt = process_options(["-c", "1 + 2"])
opt.fine_grained_incremental = True

result = build(files, options=opt)


class Visitor(TraverserVisitor):
    def visit_op_expr(self, o: OpExpr) -> None:
        super().visit_op_expr(o)

        match o:
            case OpExpr(
                left=IntExpr(value=1),
                op="+",
                right=IntExpr(value=2),
            ):
                print("found")

            case _:
                print("not found")


tree = result.graph['__main__'].tree

if tree:
    tree.accept(Visitor())
```
</details>

### PR DESCRIPTION
**Pull request by dosisod**
Closes #13243.

For the most part, in order to determine which fields where worth pattern matching against, I looked at what was being displayed in the `StrConv` class (defined in `mypy/strconv.py`). I might've missed some fields, but for the most part all of the important ones are there.

In addition, to make pattern matching feel more natural, I re-arranged the fields to better align with how they look in the source code. For example, when matching against `OpExpr`, you would do:

```python
match o:
    case OpExpr(IntExpr(1), "+", IntExpr(2)):
        pass
```

instead of:

```python
match o:
    case OpExpr("+", IntExpr(1), IntExpr(2)):
        pass
```

~~Also, deleted an old semantic analyzer-related field (`imported_names`, line 481). Didn't seem to be in use anymore.~~

### CODE REVIEW

File: mypy/nodes.py [hunk 1]
```diff
@@ -465,18 +473,17 @@ def accept(self, visitor: StatementVisitor[T]) -> T:
...
     """from m import *"""
 
-    __slots__ = ("id", "relative", "imported_names")
+    __slots__ = ("id", "relative")
+
+    __match_args__ = ("id", "relative")
 
     id: str
     relative: int
-    # NOTE: Only filled and used by old semantic analyzer.
```
Comment (sobolevn) [inline mypy/nodes.py:472]: I think that removing old parts can go into a separate PR. It is not related to `__match_args__`.

**Review by sobolevn — COMMENTED** [review]
Thank you! I like the idea, this might be useful for plugin authors.
I am going to leave verifing the contents of  `__match_args__` to someone who knows more about pattern matching than I do 🙂

**Comment by dosisod** [comment]
@sobolevn , do you know someone who would be able to review this PR? Thanks!

**Comment by sobolevn** [comment]
@JukkaL @ilevkivskyi can you please take a look? 🙏

File: mypy/nodes.py [hunk 2]
```diff
@@ -1931,6 +2017,8 @@ class OpExpr(Expression):
 
     __slots__ = ("op", "left", "right", "method_type", "right_always", "right_unreachable")
 
+    __match_args__ = ("left", "op", "left", "method_type")
```
Comment (ilevkivskyi) [inline mypy/nodes.py:2020]: Duplicate "left", should be "right"?

File: mypy/nodes.py [hunk 3]
```diff
@@ -249,6 +249,8 @@ class SymbolNode(Node):
 
     __slots__ = ()
 
+    __match_args__ = ("name", "fullname")
+
```
Comment (ilevkivskyi) [inline mypy/nodes.py:253]: I think "non-leaf" classes should not define this. Normally they never directly appear in an AST. There are also some examples below.

File: mypy/nodes.py [hunk 4]
```diff
@@ -1872,6 +1952,8 @@ class IndexExpr(Expression):
 
     __slots__ = ("base", "index", "method_type", "analyzed")
 
+    __match_args__ = ("base", "index", "method_type")
```
Comment (ilevkivskyi) [inline mypy/nodes.py:1955]: "method_type" is kind of an internal thing, it probably should not be exposed.

File: mypy/nodes.py [hunk 5]
```diff
@@ -1896,6 +1978,8 @@ class UnaryExpr(Expression):
 
     __slots__ = ("op", "expr", "method_type")
 
+    __match_args__ = ("op", "expr", "method_type")
```
Comment (ilevkivskyi) [inline mypy/nodes.py:1981]: Also remove "method_type" here an below.

File: mypy/nodes.py [hunk 6]
```diff
@@ -2542,6 +2676,8 @@ class PromoteExpr(Expression):
 
     __slots__ = ("type",)
 
+    __match_args__ = ("type",)
```
Comment (ilevkivskyi) [inline mypy/nodes.py:2679]: `PromoteExpr` is for internal use, don't think we want to expose it.

File: mypy/nodes.py [hunk 7]
```diff
@@ -3283,6 +3423,8 @@ def f(x: B[T]) -> T: ...  # without T, Any would be used here
         "eager",
     )
 
+    __match_args__ = ("name", "target", "alias_tvars")
```
Comment (ilevkivskyi) [inline mypy/nodes.py:3426]: You can add "no_args" probably for consistency with `TypeAliasExpr` (or remove it there).

**Review by ilevkivskyi — COMMENTED** [review]
Did some quick pass, generally looks good, but there are several comments.

File: mypy/nodes.py (hunk 3 above)
Comment (dosisod) [reply mypy/nodes.py]: Ok, I removed the `__match_args__` from this class. Are there any others that I should remove it from?

File: mypy/nodes.py (hunk 2 above)
Comment (dosisod) [reply mypy/nodes.py]: Thanks, I would've missed that!

File: mypy/nodes.py [hunk 8]
```diff
@@ -547,6 +555,8 @@ class FuncBase(Node):
         "_fullname",
     )
 
+    __match_args__ = ("name", "type", "info")
```
Comment (ilevkivskyi) [inline mypy/nodes.py:558]: This is a non-leaf class.

File: mypy/nodes.py [hunk 9]
```diff
@@ -713,6 +725,8 @@ class FuncItem(FuncBase):
 
     __deletable__ = ("arguments", "max_pos", "min_args")
 
+    __match_args__ = ("arguments", "type", "body")
```
Comment (ilevkivskyi) [inline mypy/nodes.py:728]: This is a non-leaf class.

**Review by ilevkivskyi — COMMENTED** [review]
There are two more non-leaf classes. Also it looks like you are missing `LambdaExpr` (and the latter is missing `__slots__` probably, can be a separate PR).

**Review by ilevkivskyi — APPROVED** [review]
